Sources/NIOWebSocket/NIOWebSocketFrameAggregator.swift

```diff
@@ -0,0 +1,142 @@
+//
+//===----------------------------------------------------------------------===//
+import NIO
+
+
+/// `NIOWebSocketFrameAggregator` buffers inbound fragmented `WebSocketFrame`'s and aggregates them into a single `WebSocketFrame`.
+/// It guarantees that a `WebSocketFrame` with an `opcode` of `.continuation` is never forwarded.
+/// Frames which are not fragmented are just forwarded without any processing.
+/// Fragmented frames are unmasked, concatenated and forwarded as a new `WebSocketFrame` which is either a `.binary` or `.text` frame.
+/// `extensionData`, `rsv1`, `rsv2` and `rsv3` are lost if a frame is fragmented because they cannot be concatenated.
+/// - Note: `.ping`, `.pong`, `.closeConnection` frames are forwarded during frame aggregation
+public final class NIOWebSocketFrameAggregator: ChannelInboundHandler {
+    public enum Error: Swift.Error {
+        case nonFinalFragmentSizeIsTooSmall
+        case tooManyFragments
+        case accumulatedFrameSizeIsTooLarge
+        case receivedNewFrameWithoutFinishingPrevious
+        case didReceiveFragmentBeforeReceivingTextOrBinaryFrame
+    }
+    public typealias InboundIn = WebSocketFrame
+    public typealias InboundOut = WebSocketFrame
+    
+    private let minNonFinalFragmentSize: Int
+    private let maxAccumulatedFrameCount: Int
+    private let maxAccumulatedFrameSize: Int
+    
+    private var bufferedFrames: [WebSocketFrame] = []
+    private var accumulatedFrameSize: Int = 0
+    
+    
+    /// Configures a `NIOWebSocketFrameAggregator`.
+    /// - Parameters:
+    ///   - minNonFinalFragmentSize: Minimum size in bytes of a fragment which is not the last fragment of a complete frame. Used to defend agains many really small payloads.
+    ///   - maxAccumulatedFrameCount: Maximum number of fragments which are allowed to result in a complete frame.
+    ///   - maxAccumulatedFrameSize: Maximum accumulated size in bytes of buffered fragments. It is essentially the maximum allowed size of an incoming frame after all fragments are concatenated.
+    public init(
+        minNonFinalFragmentSize: Int,
+        maxAccumulatedFrameCount: Int,
+        maxAccumulatedFrameSize: Int
+    ) {
+        self.minNonFinalFragmentSize = minNonFinalFragmentSize
+        self.maxAccumulatedFrameCount = maxAccumulatedFrameCount
+        self.maxAccumulatedFrameSize = maxAccumulatedFrameSize
+    }
+    
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let frame = unwrapInboundIn(data)
+        do {
+            switch frame.opcode {
+            case .continuation:
+                guard let firstFrame = self.bufferedFrames.first else {
+                    throw Error.didReceiveFragmentBeforeReceivingTextOrBinaryFrame
+                }
+                try self.bufferFrame(frame)
+                
+                guard frame.fin else { break }
+                // final frame received
+                let aggregatedFrame = firstFrame.aggregated(
```

Comment:
Hmm, sorry, now that I see this I think this will trigger a copy-on-write.

Fixing the CoW will be very fiddly so it might just be best to go back to the previous design but use the `context.channel.allocator` to get the ByteBuffer. Sorry for wasting your time there!

Reply:
Don't worry :)